rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,208 @@
+for providers. This would allow us to implement the control server as a
+provider with fewer structural changes to Vector's main run loop. Concretely,
+this could look something like adding an optional oneshot response channel to
+the `SignalTo::ReloadFromConfigBuilder` variant used by providers.
+
+We don't take this route because the signal-based focus of reloading is already
+overloaded enough to make the process difficult to follow. It's not a good
+match for what we're trying to do and would like further obscure the logic,
+making bugs more difficult to track down. In fact, given the limited uptake of
+the provider system in general, it's possible that it should be reworked to fit
+the same pattern as the new control server. This could be done as future work.
+
+### More isolated control socket mode
+
+Another option would be to branch earlier in `Application::run`, make options
+related to the control server exclusive with normal options like `--config`,
+and have more of a separate mode of operation for this functionality. This
+would remove some concerns about potential strange interactions between
+multiple paths for reloading (e.g. watching a file while running a control
+server while running an HTTP provider) and potentially require less refactoring
+work than what's proposed in this RFC.
+
+This approach wasn't chosen because the divergence for "normal" Vector
+operation would itself be a path for potentially confusing differences in
+behavior. It also would not move us closer to an improved overall design for
+this portion of Vector, and would create another path that we would later
+likely want to unify.
+
+## Outstanding Questions
+
+- What is the expected flow for setting global options at startup (especially
+  those that can't be changed via reload)? Use normal file loading with only
+  global options set?
+- Should the API accept entire configs, or should we attempt to limit to just
+  sources/transforms/sinks?
+
+## Plan Of Attack
+
+- [ ] Extract controller struct to consolidate things that change with reloads
+- [ ] Write simple Unix socket HTTP server task with `POST /config` route that builds/validates config
+- [ ] Wrap and share controller struct with server task so it can actually do reloads
+- [ ] Refactor topology reload method to return errors rather than log directly
+
+## Future Improvements
+
+- Write an RFC for differentiating different levels of configuration (e.g.
+  global vs component, reloadable vs not) and modeling them better within
+  Vector. Then use this to have a clearer distinction between what should be
+  present in a global config file vs what's loadable via the API, make things
+  appropriately exclusive and reduce the number of possible weird interactions
+  from running multiple config loading mechanisms concurrently.
+- Migrate features like `top` and `tap` to use the new API and reduce
+  maintenance cost of old + new API.
```

Comment:
Agreed. I'd like to see some more discussion of how we'll replace the GraphQL API before adding a new mechanism to avoid maintaining both. Left a similar comment here: https://github.com/vectordotdev/vector/pull/15562#issuecomment-1357888941